File: Data Structures & Algorithms/reorder-linked-list/submission-4.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public void reorderList(ListNode head) {
        if(head==null || head.next==null ) return ;
        ListNode slow=head;
        ListNode fast=head.next;

        while(fast!=null && fast.next!=null)
        {
            fast=fast.next.next;
            slow=slow.next;
        }


        ListNode prev=null;
        ListNode next=null;
        ListNode curr=slow.next;
        while(curr!=null)
        {
            next=curr.next;
            curr.next=prev;
            prev=curr;
            curr=next;
        }

        slow.next=null;

        ListNode temp=head;

        while(prev!=null)
        {
            ListNode tmp1=temp.next;
            ListNode tmp2=prev.next;
            temp.next=prev;
            prev.next=tmp1;
            prev=tmp2;
            temp=tmp1;
        }
        

    }
}

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-4.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        
        curr= head
        
        t = 0

        temp = head
        while temp:
            t+=1
            temp=temp.next

        print(t)
        n = t-n
        if n == 0:
            return head.next
        while n>1:
            curr = curr.next
            n-=1
        temp = curr.next
        if temp:
            temp = temp.next
            curr.next = temp
        else:
            curr.next = None

        return head
        

